DNM: testing #144 with temporary replace lines

gibizer opened the pull request:
Just testing #144

SeanMooney commented:
this is kind of hard to review since you have a mix fo the autogenerated code form the SDK and the actual functional change in the same commit
it would be better to do this as two commits

SeanMooney commented on config/default/webhookcainjection_patch.yaml:
```diff
@@ -0,0 +1,29 @@
+    app.kubernetes.io/managed-by: kustomize
+  name: mutating-webhook-configuration
+  annotations:
+    cert-manager.io/inject-ca-from: $(CERTIFICATE_NAMESPACE)/$(CERTIFICATE_NAME)
```
this comes form teh sdk but we should not have the cert manager dep since that is tech preview and we wont be able to use it in production unless that is GA'd before we release this.

dprince replied:
@SeanMooney we have prior art for this in multiple other operators. I think we have agreed that for now we are going with it.

dprince replied:
we already require this in openstack-operator, and glance-operator. I think the agreement is that we accept the risk and want to use it for the time being

dprince replied:
both openstack-operator and glance-operator already use similar code for webhook installation

gibizer commented on api/v1beta1/placementapi_types.go:
```diff
@@ -53,8 +53,7 @@ type PlacementAPISpec struct {
-	// +kubebuilder:validation:Optional
-	// +kubebuilder:default="quay.io/tripleozedcentos9/openstack-placement-api:current-tripleo"
+	// +kubebuilder:validation:Required
 	// PlacementAPI Container Image URL
```
I would describe here that if this is not provided then it will be defaulted automatically based on the environment of the placement-operator. (Or something similar so that the end user looking at the description can realize that she does not need to provide any specific value here by default)

gibizer commented on api/v1beta1/placementapi_webhook.go:
```diff
@@ -0,0 +1,98 @@
+var placementapilog = logf.Log.WithName("placementapi-resource")
+
+// SetupDefaults - initialize PlacementAPI spec defaults for use with either internal or external webhooks
+func (spec *PlacementAPISpec) SetupDefaults(defaults PlacementAPIDefaults) {
```
`SetupDefaults` does not need to have `PlacementAPISpec` receiver. It only sets the global `placementAPIDefaults` so it can be global

abays replied:
I do like that it makes it obvious as to what the defaults pertain.  That being said, I suppose a global function named `SetupPlacementAPIDefaults` or something similar would provide the same clarity.

gibizer replied:
Lets go with `SetupPlacementAPIDefaults` :)

gibizer commented on api/v1beta1/placementapi_webhook.go:
```diff
@@ -0,0 +1,98 @@
+
+// Default - set defaults for this PlacementAPI spec
+func (spec *PlacementAPISpec) Default() {
+	if spec.ContainerImage == "" {
```
(Just a note, nothing to change here.) This means PlacementAPI.Spec.ContainerImage cannot ever be set to "" explicitly as that will always be overridden by the defaulting here. We need to be careful and conscious when making the decision that the golang default value of a type is used as the magic value for unset. For pointer types `nil` is good for that but for non pointer types it can be dangerous. E.g. for bool with a default true, we cannot use == false to detect the unset state.

abays replied:
Ack

gibizer commented on api/v1beta1/placementapi_webhook.go:
```diff
@@ -0,0 +1,98 @@
+var placementAPIDefaults PlacementAPIDefaults
+
+// log is for logging in this package.
+var placementapilog = logf.Log.WithName("placementapi-resource")
```
just out of curiosity. Where this log will be emitted in a real k8s cluster / CRC? Is it still part of the log of the operator-controller-manager pod?

abays replied:
Yes, I believe so

gibizer replied:
cool thanks. I'm wondering how can we make this more aligned with the structural logging we have in the reconciler but the framework does not really expose as similar ctx than the Reconciler gets. So this I guess this is OK as is for now.

gibizer commented on api/v1beta1/placementapi_webhook.go:
```diff
@@ -0,0 +1,98 @@
+
+	// TODO(user): fill in your validation logic upon object deletion.
+	return nil
+}
```
I think we eventually want to have validation functions on the PlacementAPISpec level to be called from the openstack-operator as that operator only have PlacementAPISpec object not a full PlacementAPI object. I think we need a similar two level indirection as you proposed for the Defaulter.

abays replied:
Yes, that is the plan and has already been implemented for OpenStack operator [1] and Ironic [2] as a first instantiation.

[1] https://github.com/openstack-k8s-operators/openstack-operator/blob/5257475765050d0be683c4cd9e04ad68f307cede/apis/core/v1beta1/openstackcontrolplane_webhook.go#L162-L167
[2] https://github.com/openstack-k8s-operators/ironic-operator/blob/deae2b4e4e8b5ef12f940590aae68d10181de88f/api/v1beta1/ironic_webhook.go#L70-L110

abays commented on api/v1beta1/placementapi_webhook.go:
```diff
@@ -0,0 +1,98 @@
+
+// SetupDefaults - initialize PlacementAPI spec defaults for use with either internal or external webhooks
+func (spec *PlacementAPISpec) SetupDefaults(defaults PlacementAPIDefaults) {
+	placementAPIDefaults = defaults
```
Ack

abays commented on api/v1beta1/webhook_suite_test.go:
```diff
@@ -0,0 +1,31 @@
+
+func TestAPIs(t *testing.T) {
+	RegisterFailHandler(Fail)
+}
```
I will try to get these scaffolded tests working in the next operator in which I add webhooks.  They do not work out-of-the-box and I need to dig into the failures.

abays commented on main.go:
```diff
@@ -110,6 +111,22 @@ func main() {
+		ContainerImageURL: os.Getenv("PLACEMENT_API_IMAGE_URL_DEFAULT"),
+	}
+
+	(&placementv1.PlacementAPI{}).Spec.SetupDefaults(placementAPIDefaults)
```
Ack

abays commented on hack/configure_local_webhook.sh:
```diff
@@ -0,0 +1,89 @@
+FIREWALL_ZONE=${FIREWALL_ZONE:-"libvirt"}
+
+#Open 9443
+sudo firewall-cmd --zone=${FIREWALL_ZONE} --add-port=9443/tcp
+sudo firewall-cmd --runtime-to-permanent
```
If we're going to use webhooks, I think it's better that we provide something rather than leave devs to figure out how to run webhooks locally themselves (even if this script won't work exactly as-is for them given their particular environment).  You can also just disable the webhooks locally if you don't actually need what they're doing for your dev purposes (i.e. `ENABLE_WEBHOOKS=false ... make run`).  So far I have used this script against an `install_yamls` environment on my laptop and also with an `osp-director-dev-tools` full OCP cluster on a dedicated server.

gibizer replied:
I agree that having something is better than nothing. Let's table this comment until I have time to play with this locally. I expect I will need to push a follow up to make this working for me as well. But that is probably fine. If the this works for the majority of the devs, then I will adapt.

gibizer commented on api/v1beta1/placementapi_webhook.go:
```diff
@@ -0,0 +1,98 @@
+}
+
+// Default - set defaults for this PlacementAPI spec
+func (spec *PlacementAPISpec) Default() {
```
OK. cool. Thanks for checking it.

stuggi reviewed (commented):
looks good to me @gibizer are your questions/comments addressed and we can land this?

gibizer reviewed (approved):
looks good thanks

Commit:
Update controller-runtime version

Signed-off-by: Brendan Shephard <bshephar@redhat.com>

Commit:
[webhook] Update webhook functions for CR update

This change updates the webhook functions to comply with the new
standard for controller-runtime.

Signed-off-by: Brendan Shephard <bshephar@redhat.com>

Commit:
[manager] Update manager initialisation for CR update

This change updates the manager intialisation to comply with the new controller-runtime schema.

Signed-off-by: Brendan Shephard <bshephar@redhat.com>

Commit:
[watches] Fix watches declarations for CR update

This change fixes the Watches declarations to comply with the new
controller-runtime standards

Signed-off-by: Brendan Shephard <bshephar@redhat.com>

Commit:
Re-render CRD manifests

Signed-off-by: Brendan Shephard <bshephar@redhat.com>

Commit:
Update to Go 1.21

Signed-off-by: Brendan Shephard <bshephar@redhat.com>

gibizer committed:
DNM: test with replace lines

gibizer commented:
Locally the functional test crashes
```
Running Suite: Controller Suite - /tmp/work/tests/functional
============================================================
Random Seed: 1707323975 - will randomize all specs

Will run 29 of 29 specs
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x189faa9]

goroutine 456 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.0/pkg/internal/controller/controller.go:116 +0x1e5
panic({0x1ac19c0?, 0x2f1f430?})
	/usr/local/go/src/runtime/panic.go:914 +0x21f
github.com/openstack-k8s-operators/mariadb-operator/api/v1beta1.(*MariaDBAccount).GetNamespace(0x1d54012?)
	<autogenerated>:1 +0x9
github.com/openstack-k8s-operators/lib-common/modules/common/util.logObjectParams({0x2050088?, 0x0})
	/go/pkg/mod/github.com/bshephar/lib-common/modules/common@v0.0.0-20240205000728-69fddd97f72e/util/log.go:30 +0x89
github.com/openstack-k8s-operators/lib-common/modules/common/util.LogForObject(0xc00056ac60, {0xc000240d20, 0x58}, {0x2050088?, 0x0?}, {0x0, 0x0, 0x0})
	/go/pkg/mod/github.com/bshephar/lib-common/modules/common@v0.0.0-20240205000728-69fddd97f72e/util/log.go:42 +0x54
github.com/openstack-k8s-operators/mariadb-operator/api/v1beta1.(*Database).getDBWithName(0xc00067b908, {0x203ac80, 0xc000bda000}, 0xc00056ac60)
	/go/pkg/mod/github.com/bshephar/mariadb-operator/api@v0.0.0-20240206224814-4264f1e7465a/v1beta1/mariadbdatabase_funcs.go:401 +0x3e5
github.com/openstack-k8s-operators/mariadb-operator/api/v1beta1.(*Database).CreateOrPatchDBByName(0xc00067b908, {0x203ac80, 0xc000bda000}, 0xc00056ac60, {0x1d5ac7e, 0x9})
	/go/pkg/mod/github.com/bshephar/mariadb-operator/api@v0.0.0-20240206224814-4264f1e7465a/v1beta1/mariadbdatabase_funcs.go:257 +0x6a6
github.com/openstack-k8s-operators/mariadb-operator/api/v1beta1.(*Database).CreateOrPatchDB(...)
	/go/pkg/mod/github.com/bshephar/mariadb-operator/api@v0.0.0-20240206224814-4264f1e7465a/v1beta1/mariadbdatabase_funcs.go:139
github.com/openstack-k8s-operators/placement-operator/controllers.(*PlacementAPIReconciler).reconcileInit(0xc000720540?, {0x203ac80, 0xc000bda000}, 0xc000239180, 0x0?, 0xc000a8a810, 0xc000c23bf0?)
	/tmp/work/controllers/placementapi_controller.go:541 +0x5a5
github.com/openstack-k8s-operators/placement-operator/controllers.(*PlacementAPIReconciler).reconcileNormal(0x2052248?, {0x203ac80, 0xc000bda000}, 0xc000239180, 0xc00056ac60)
	/tmp/work/controllers/placementapi_controller.go:980 +0x150b
github.com/openstack-k8s-operators/placement-operator/controllers.(*PlacementAPIReconciler).Reconcile(0xc000744ea0, {0x203ac80?, 0xc000bda000}, {{{0xc00054cb10?, 0x5?}, {0xc000992bc0?, 0xc00067dd08?}}})
	/tmp/work/controllers/placementapi_controller.go:322 +0x1036
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x203eb60?, {0x203ac80?, 0xc000bda000?}, {{{0xc00054cb10?, 0xb?}, {0xc000992bc0?, 0x0?}}})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.0/pkg/internal/controller/controller.go:119 +0xb7
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0004f2000, {0x203acb8, 0xc00045ce10}, {0x1b7f8c0?, 0xc000249ea0?})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.0/pkg/internal/controller/controller.go:316 +0x3cc
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0004f2000, {0x203acb8, 0xc00045ce10})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.0/pkg/internal/controller/controller.go:266 +0x1af
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.0/pkg/internal/controller/controller.go:227 +0x79
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2 in goroutine 208
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.0/pkg/internal/controller/controller.go:223 +0x565
ginkgo run failed
  could not finalize profiles:
  Unable to read coverage file /tmp/work/tests/functional/cover.out:
  open /tmp/work/tests/functional/cover.out: no such file or directory
make: *** [Makefile:112: test] Error 1
```

I think what is happening that the mariadb-operator/api bump brings in new commits from mariadb-operator that is not compatible with the current placement-operator code. We probably need separate PR to bump to the latest merged mariadb-operator/api first and fix things there.

gibizer commented:
https://github.com/openstack-k8s-operators/placement-operator/pull/146 This shows that the crash is present without the controller-runtime bump and only need to reproduce it is mariadb bump
